### PR TITLE
Reimplement Jsonify type utility

### LINKED ISSUE #7246: V2 issue: inferred type disappearing
### What version of Remix are you using?

0.0.0-nightly-49e8da1-20230823

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

```tsx
import * as React from "react";
import { json } from "@remix-run/node"; // types: 0.0.0-nightly-49e8da1-20230823
import { useLoaderData } from "@remix-run/react"; // types: 0.0.0-nightly-49e8da1-20230823

export async function loader() {
  return json({ payload: { name: "bob" } as Record<string, any> });
}

export default function RouteComponent() {
  const data = useLoaderData<typeof loader>();

  console.log(data.payload); // <-- Property 'payload' does not exist on type 'JsonifyObject<{}>'.ts(2339)

  return <div>stuff</div>;
}
```

I can't get this to reproduce in the TypeScript Playground for some reason, but [here's that](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQokscA3nAFaoQB2cAvnCIlyAAUohgADwC0UAK7cA9NwgATLLTiLFcGAE8wWVAC44ABgB0ls9O7AA5gAsYAGz3SALAE4sADlXIAIzSAExmIQDMZr6RDODQ8GxyqFgAMhDI6lAAIsgwyPyCxKRkYlgSMvJKlHgwmtq6Bkam1la2Ds5unj7+QaHhUTERdHRYkkzwaHrcuIIK+MA8cC4ZWQAUAJSsdHBwlDByULycPGtsYMh6K5mmbNzIIFimZABGEC9khWiYWLjQqgAeVAwKDAbj2AA0KG4egAfPwNvQ+CMxhM4OoCMg5C5CPMYIteBgIHIYFgAMIkSDcLDcGCbba7P7cYHovIFAC8cGSaVWWBybIB+kMEAIy15UFhm3oOzgTK4LiwFhW9jWAXyFguV1WiJGu32h14ANUwAAbrDgXICAQAYpjWakXQgA)

Interestingly there's a difference in the type on the TS Playground and my local version:

image

image

I also tried this out on a fresh project, so I'm confident it's not just my project.

### Expected Behavior

I expect to be able to infer types of `Record<string, any>`

### Actual Behavior

That property disappears.

### PR DESCRIPTION
**Pull request by pcattori**
Fixes #7246 , #3931

`type-fest` implementation is causing type errors for Remix users. Tried fixing the errors in `type-fest`, but the repo setup causes some errors not to be reproducible there (and to be inconsistent with Typescript playgrounds). Additionally, there were many test cases in `type-fest` for `Jsonify` whose "expected vs actual" I disagreed with.

So for now, we can reintroduce our own implementation. Later, `type-fest` can choose which parts they wish to adopt or not.

## Testing strategy

Tests are collocated in the same file as the implementation and are typecheck-only tests.

## TODO

- [x] Implement @nrako 's suggestion
- [x] changeset
- [ ] later: write up a doc to address misunderstandings from #3931

### CODE REVIEW

**Comment by nrako** [comment]
@pcattori - I've recently noticed your PR, and I indeed believe that it makes sense to repatriate this logic back to Remix, especially since the adoption of `type-fest` seems to be causing significant migration issues for several Remix's users. Moreover, this feature, in my humble opinion, is quite crucial and central to Remix, and since only `Jsonify` is used, I think your changes are a very sensible approach for Remix.

However, I'd like to bring to your attention the following PR I had submitted for type-fest: https://github.com/sindresorhus/type-fest/pull/690

This PR addresses an issue with `type-fest`, which I experienced through Remix v2, and seeing your PR, I'm wondering if you could incorporate my patch into your own PR.

**Comment by nrako** [comment]
To help you understand the gist of the problem, please find bellow an example of a patch:

```diff
diff --git a/packages/remix-server-runtime/jsonify.ts b/packages/remix-server-runtime/jsonify.ts
index 9531f3d51..f78b2d063 100644
--- a/packages/remix-server-runtime/jsonify.ts
+++ b/packages/remix-server-runtime/jsonify.ts
@@ -5,6 +5,9 @@ export type Jsonify<T> =
   // any
   IsAny<T> extends true ? any :
 
+  // toJSON which is tested first in case an object also implement a "JSONPrimitive" interface
+  T extends { toJSON(): infer U } ? (U extends JsonValue ? U : unknown) :
+
   // primitives
   T extends JsonPrimitive ? T :
   T extends String ? string :
@@ -21,9 +24,6 @@ export type Jsonify<T> =
   // Not JSON serializable
   T extends NotJson ? never :
 
-  // toJSON
-  T extends { toJSON(): infer U } ? (U extends JsonValue ? U : unknown) :
-
   // tuple & array
   T extends [] ? [] :
   T extends readonly [infer F, ...infer R] ? [NeverToNull<Jsonify<F>>, ...Jsonify<R>] :
@@ -141,6 +141,8 @@ export type _tests = [
   Expect<Equal<Jsonify<{ toJSON(): undefined }>, unknown>>,
   Expect<Equal<Jsonify<{ toJSON(): Date }>, unknown>>,
 
+  Expect<Equal<Jsonify<BooleanWithToJson>, string>>,
+
   // tuple & array
   Expect<Equal<Jsonify<[]>, []>>,
   Expect<Equal<Jsonify<[1, 'two', Date, undefined, false]>, [1, 'two', string, null, false]>>,
@@ -244,3 +246,7 @@ type EmptyObject = { [emptyObjectSymbol]?: never };
 
 // adapted from https://github.com/type-challenges/type-challenges/blob/main/utils/index.d.ts
 type IsAny<T> = 0 extends 1 & T ? true : false;
+
+interface BooleanWithToJson extends Boolean {
+  toJSON(): string;
+}

```

p.s: I like your approach of simply collocating the typecheck-only tests in the file directly 👍🏽

**Comment by pcattori** [comment]
@nrako thanks for the suggestion! I included your improvement and added you as a co-author for that commit

File: packages/remix-server-runtime/jsonify.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+  // TypedArray
+  T extends TypedArray ? Record<string, number> :
+
+  // Not JSON serializable
+  T extends NotJson ? never :
+
+  // tuple & array
+  T extends [] ? [] :
+  T extends readonly [infer F, ...infer R] ? [NeverToNull<Jsonify<F>>, ...Jsonify<R>] :
+  T extends readonly unknown[] ? Array<NeverToNull<Jsonify<T[number]>>>:
```
Comment (pcattori) [inline packages/remix-server-runtime/jsonify.ts:37]:
```suggestion
  T extends readonly Array<infer Item>? Array<NeverToNull<Jsonify<Item>>>:
```

**Review by markdalgleish — APPROVED** [review]

File: packages/remix-server-runtime/serialize.ts [hunk 2]
```diff
@@ -1,10 +1,7 @@
-import type { Jsonify } from "type-fest";
```
Comment (pcattori) [inline packages/remix-server-runtime/serialize.ts:1]: Are we still using `type-fest` elsewhere? or can we get rid of that dep?

File: packages/remix-server-runtime/serialize.ts [hunk 3]
```diff
@@ -13,29 +10,74 @@ type Fn = (...args: any[]) => any;
...
-    Jsonify<Awaited<Output>> :
+export type SerializeFrom<T> =
+  T extends (...args: any[]) => infer Output ? Serialize<Awaited<Output>> :
+  // Back compat: manually defined data type, not inferred from loader nor action
   Jsonify<Awaited<T>>
 ;
 
+// note: cannot be inlined as logic requires union distribution
+type Serialize<Output> = Output extends TypedDeferredData<infer U>
+  ? // top-level promises
```
Comment (pcattori) [inline packages/remix-server-runtime/serialize.ts:21]: format this better

File: packages/remix-server-runtime/typecheck.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// typecheck that expression is assignable to type
+
+export function expectType<T>(_expression: T) {}
```
Comment (pcattori) [inline packages/remix-server-runtime/typecheck.ts:3]: add line after this

**Pull request merged**